src/Orleans.Core/Serialization/TypeUtilities.cs

```diff
@@ -21,86 +20,76 @@ internal static bool IsOrleansPrimitive(this Type t)
-            shallowCopyableTypes[typeof(GrainId)] = true;
-            shallowCopyableTypes[typeof(ActivationId)] = true;
-            shallowCopyableTypes[typeof(ActivationAddress)] = true;
-            shallowCopyableTypes[typeof(CorrelationId)] = true;
-            shallowCopyableTypes[typeof(string)] = true;
-            shallowCopyableTypes[typeof(Immutable<>)] = true;
-            shallowCopyableTypes[typeof(CancellationToken)] = true;
-        }
+            [typeof(Decimal)] = true,
+            [typeof(DateTime)] = true,
+            [typeof(TimeSpan)] = true,
+            [typeof(IPAddress)] = true,
+            [typeof(IPEndPoint)] = true,
+            [typeof(SiloAddress)] = true,
+            [typeof(GrainId)] = true,
+            [typeof(ActivationId)] = true,
+            [typeof(ActivationAddress)] = true,
+            [typeof(CorrelationId)] = true,
+            [typeof(string)] = true,
+            [typeof(CancellationToken)] = true,
+            [typeof(Guid)] = true,
+        };
 
         internal static bool IsOrleansShallowCopyable(this Type t)
         {
-            bool result;
-            if (shallowCopyableTypes.TryGetValue(t, out result))
+            if (shallowCopyableTypes.TryGetValue(t, out var result))
             {
                 return result;
             }
+            return shallowCopyableTypes.GetOrAdd(t, IsShallowCopyableInternal(t));
+        }
 
-            var typeInfo = t.GetTypeInfo();
-            if (typeInfo.IsPrimitive || typeInfo.IsEnum)
-            {
-                shallowCopyableTypes[t] = true;
+        private static bool IsShallowCopyableInternal(Type t)
+        {
+            if (t.IsPrimitive || t.IsEnum)
                 return true;
-            }
 
-            if (typeInfo.GetCustomAttributes(typeof(ImmutableAttribute), false).Any())
-            {
-                shallowCopyableTypes[t] = true;
+            if (t.GetCustomAttributes(typeof(ImmutableAttribute), false).Length != 0)
                 return true;
-            }
 
-            if (typeInfo.IsGenericType && typeInfo.GetGenericTypeDefinition() == typeof(Immutable<>))
+            if (t.IsConstructedGenericType)
             {
-                shallowCopyableTypes[t] = true;
-                return true;
+                var def = t.GetGenericTypeDefinition();
+
+                if (def == typeof(Immutable<>))
+                    return true;
+
+                if (def == typeof(Nullable<>)
+                    || def == typeof(Tuple<>)
+                    || def == typeof(Tuple<,>)
+                    || def == typeof(Tuple<,,>)
+                    || def == typeof(Tuple<,,,>)
+                    || def == typeof(Tuple<,,,,>)
+                    || def == typeof(Tuple<,,,,,>)
+                    || def == typeof(Tuple<,,,,,,>)
+                    || def == typeof(Tuple<,,,,,,,>))
+                    return Array.TrueForAll(t.GenericTypeArguments, a => IsOrleansShallowCopyable(a));
```

Comment:
Could also add ValueTuple here

Reply:
ValueTuple should be covered by the general value-type code below already.